content/wiki/features/Distributed-Content.md

```diff
@@ -0,0 +1,133 @@
+
+An example for such a list would be:
+
+- how often is the content played
+- how many people play on the _game server_ in average
+- when was the content published
+- how many storage do I have available for sharing?
+- how often would I like to propagate new metadata from the network (e.g every 5 minutes)
+- (...)
+
+Based on these parameters every node decides for themselves:
+
+- how much metadata is stored
+- how often the metadata is updated
+- which specific kind of metadata is stored
+
+This is particully usefull for specific servers, such as e.g coop servers, which do not want to store racing or competetive content (no point in updating stats every minute).
+
+### Distributing the content
+
+Now that we have the metadata settled, let's consider how the network would look like:
+
+- new players would contact the seed nodes
+- get a list of existing nodes (e.g other servers)
+- download metadata about newest maps for e.g racing
+
+But this doesn't solve the problem of storing actual content.
+
+The technical solution to this is quite straighforward: the metadata contains a hyperlink and a checksum to tell the node where to find the content.
+
+It is up to the __servers__ in the network to download the content and make it available. This can be done in exactly the same fashion as for the metadata, using the __popularity factor__.
+
+### All about authentication
+
+In the network introduced thus far, one cannot authenticate themselves. My suggestion is to use [SASL](https://en.wikipedia.org/wiki/Simple_Authentication_and_Security_Layer) with HTTPS. The authentication would work like this:
+
+- _node_ A requests authentication for _node_ B from _node_ C
+- _node_ B is redirected for log in to _node_ C
+- _node_ C asks for authentication from _node_ B
+- _node_ B authenticates
+- _node_ C verifies _node_ A that _node_ B is _B@C_
+
+(note: in this example A is a server, B is a client, and C is an __authentication provider__, where A and C could be identical)
+
+This allows for secure authentication using widely adopted standards (SASL, SSL) where implementations exist. The SSL certificates in place should come from Let's Encrypt.
+
+### Scenario: user uploads a map
+
+Ok, quite a lot of theory, but how would that look like on a practical example:
+
+- A opens his game
+- A contacts the seed nodes and gets a list of other existing nodes
+- A asks all nodes he knows about metadata based on his popularity factors
+- Since A is a client and does not share any metadata, there is no more interaction with the network
+- Now, ideally, A has a list of servers and current content
+- A decides to go into coop on server B
+- A makes a map on server B and would like to share it
+- A logs into server B and is now logged in as A@B (e.g fohlen@inexor.org)
+- A tells B that he wants the current map to be uploaded
+- B creates metadata
+- More people use the map created by A
+- It gets spread from server B to more servers
```

Comment:
I would like to still be able to choose whether or not I only want this server to keep this copy of the map or have it distributed.
This plays together with the publishing mechanisms. Do map names have to be unique then? Or only the metadata associated with it?
I think the player has to actively "publish" his content to be ready for sharing between other servers.
However the current state can always be shared, so no more `sendmap` or `getmap`.

Both local and server state should be identical.

**Also this would have many other implications:**
Players need to give other players publishing permission, otherwise any user could come, change the map and then re-publish it.

**Popularity Factor**
This would be important meta-data to be shared between nodes.
If we are able to utilize the popularity factor also for same-named content, we could easily sort contents by relevance in the asset browser/catalog

Reply:
I actually thought that would be implied but it'd be good to clarify it:

- you can decide whether or not your content will be published
- this is basically the server's popularity factor that is modified by the user? (include map: yes / no)
- no more `sendmap` or `getmap`
- for the uniqueness of content: this part of the specification is yet incomplete and we should thoroughly discuss is. the obvious solution here could be: every server has git. but this has strong implications for replication, and I believe it is not the _easy_ solution.
- sharing the popularity factor: I have not thought about this but it seems like a really good idea. We will need to provide some defaults, however it should be possible to share your config with others!